datafusion/expr/src/expr.rs

```diff
@@ -373,6 +373,20 @@ impl ScalarFunctionDefinition {
             ScalarFunctionDefinition::Name(func_name) => func_name.as_ref(),
         }
     }
+
+    /// Whether this function is volatile, i.e. whether it can return different results
+    /// when evaluated multiple times with the same input.
+    pub fn is_volatile(&self) -> bool {
+        match self {
+            ScalarFunctionDefinition::BuiltIn(fun) => {
+                fun.volatility() == crate::Volatility::Volatile
+            }
+            ScalarFunctionDefinition::UDF(udf) => {
+                udf.signature().volatility == crate::Volatility::Volatile
+            }
+            ScalarFunctionDefinition::Name(_) => false,
```

Comment:
Under which condition does the optimizer see an unresolved function name? And: if we don't know how the function behaves (because there's only a name) shouldn't we assume the worst case (i.e. that it is volatile)?

Reply:
I agree  we should assume it is volatile  given lack of additional information

At this moment, I don't think the optimizer will ever see unresolved function names

The plan is to use them to implement `expr_fn`s like [`power`](https://docs.rs/datafusion/latest/datafusion/logical_expr/fn.power.html) once we remove the hard coded list of built in functions as decsribed in https://github.com/apache/arrow-datafusion/issues/8045

In fact @edmondop  is working on this as part of https://github.com/apache/arrow-datafusion/issues/8157

Reply:
The "resolve" pass should probably run before other optimizers then, right? So maybe it's even safe to error out if we hit an unresolved function here.

Reply:
It makes sense to get error on unresolved scalar function here. Missing that it is a unresolved one. (maybe renaming it to `Unresolved` 🤔 ).